packages/server/test/links.test.ts

```diff
@@ -0,0 +1,390 @@
+        path: 'hello',
+        input: null,
+      },
+    });
+
+    const $result2 = executeChain({
+      links,
+      op: {
+        type: 'query',
+        path: 'hello',
+        input: 'alexdotjs',
+      },
+    });
+
+    await waitFor(() => {
+      expect($result1.get()).not.toBeNull();
+      expect($result2.get()).not.toBeNull();
+    });
+    expect($result1.get()).toMatchObject({
+      data: 'hello world',
+    });
+    expect($result2.get()).toMatchObject({
+      data: 'hello alexdotjs',
+    });
+
+    expect(contextCall).toHaveBeenCalledTimes(1);
+
+    close();
+  });
+
+  test('server not configured for batching', async () => {
+    const serverCall = jest.fn();
+    const { close, router, port, trpcClientOptions } = routerToServerAndClient(
+      trpc.router().query('hello', {
+        resolve() {
+          serverCall();
+          return 'world';
+        },
+      }),
+      {
+        server: {
+          batching: {
+            enabled: false,
+          },
+        },
+      },
+    );
+    const client = createTRPCClient<typeof router>({
+      ...trpcClientOptions,
+      links: [
+        httpBatchLink({
+          url: `http://localhost:${port}`,
+        }),
+      ],
+      headers: {},
+    });
+
+    await expect(client.query('hello')).rejects.toMatchInlineSnapshot(
+      `[Error: Batching is not enabled on the server]`,
+    );
+
+    close();
+  });
+});
+
+test('split link', () => {
+  const left = jest.fn();
+  const right = jest.fn();
+  executeChain({
+    links: [
+      splitLink({
+        left: () => left,
+        right: () => right,
+        condition(op) {
+          return op.type === 'query';
+        },
+      })(mockRuntime),
+    ],
+    op: {
+      type: 'query',
+      input: null,
+      path: '',
+    },
+  });
+  expect(left).toHaveBeenCalledTimes(1);
+  expect(right).toHaveBeenCalledTimes(0);
+});
+
+test('create client with links', async () => {
+  let attempt = 0;
+  const serverCall = jest.fn();
+  const { close, router, port, trpcClientOptions } = routerToServerAndClient(
+    trpc.router().query('hello', {
+      resolve() {
+        attempt++;
+        serverCall();
+        if (attempt < 3) {
+          throw new Error('Errr ' + attempt);
+        }
+        return 'world';
+      },
+    }),
+  );
+  const client = createTRPCClient<typeof router>({
+    ...trpcClientOptions,
+    links: [
+      retryLink({ attempts: 3 }),
+      httpLink({
+        url: `http://localhost:${port}`,
+      }),
+    ],
+    headers: {},
+  });
+
+  const $result = await client.query('hello');
+  expect($result).toBe('world');
+
+  close();
+});
+
+test('multi down link', async () => {
+  const $result = executeChain({
+    links: [
+      // mock cache link
+      ({ prev, onDestroy }) => {
+        const timer = setTimeout(() => {
+          prev({ ok: true, data: 'cached2', statusCode: 200 });
+        }, 1);
+        onDestroy(() => {
+          clearTimeout(timer);
+        });
```

Comment:
cc @SimonRichardson  -- `onDestroy` is called if the caller calls `$result.abort()`

Reply:
(maybe it should be called `onAbort`?)